[WIP] Add jsonparser example

DanielBicek opened the pull request:
Adds a json parser as an example. Not yet fully compliant with this https://github.com/nst/JSONTestSuite JSON test suite.

DanielBicek committed:
Add jsonparser example

timsueberkrueb committed:
Fix target folder not being ignored

timsueberkrueb commented:
Failing tests:
* [x] `n_multidigit_number_then_00` (probably) because of Logos interpreting `\u0000` as EOI
* [x] `n_string_escape_x` [Regex](https://regexr.com/44jvi) seems to be fine (fixed in Logos 0.9)
* [x] `n_string_escaped_emoji` [Regex](https://regexr.com/44jvu) seems to be fine (fixed in Logos 0.9)
* [x] `n_string_invalid_backslash_esc` [Regex](https://regexr.com/44k04) seems to be fine (fixed in Logos 0.9)
* [x] `n_string_incomplete_surrogate_escape_invalid`: [Regex](regexr.com/44k07) seems to be  (fixed in Logos 0.9)fine, result: `Ok(Array([String("\\uD800\\uD800\\x")]))`
* [x] `n_string_unescaped_newline` Fix the test case and add `\n` to the negated set
* [x] `n_string_unicode_CapitalU` [Regex](https://regexr.com/44ju2) [should be correct](https://paste.ubuntu.com/p/tC2Nkgnfrb/). Still, Logos parses the input: (fixed in Logos 0.9)
* [x] `n_string_unescaped_tab` Changing the negated set to include `\t` (like so: `[^"\\\t]`) [fixes the issue](https://regexr.com/44juh)
* [x] `n_structure_whitespace_formfeed` `\f` is (probably) allowed because there is currently no way in Logos to select which kinds of whitespace should the recognized: https://github.com/maciejhirsz/logos/issues/21

DanielBicek committed:
Fix test case naming and order tests by expected result

timsueberkrueb committed:
Fix n_string_unescaped_tab

DanielBicek committed:
fix n_string_unescaped_newline

timsueberkrueb commented:
The following tests from the suite are currently ignored:
```
n_array_a_invalid_utf8.json
i_string_iso_latin_1.json
i_string_overlong_sequence_6_bytes_null.json
n_number_invalid-utf-8-in-exponent.json
n_structure_single_eacute.json
n_number_invalid-utf-8-in-int.json
n_structure_lone-invalid-utf-8.json
n_number_invalid-utf-8-in-bigger-int.json
n_object_lone_continuation_byte_in_key_and_trailing_comma.json
n_number_real_with_invalid_utf8_after_e.json
i_string_overlong_sequence_2_bytes.json
i_string_truncated-utf-8.json
i_string_UTF-16LE_with_BOM.json
i_string_UTF-8_invalid_sequence.json
i_string_not_in_unicode_range.json
i_string_lone_utf8_continuation_byte.json
i_string_utf16LE_no_BOM.json
i_string_invalid_utf-8.json
n_array_invalid_utf8.json
n_string_invalid_utf8_after_escape.json
i_string_utf16BE_no_BOM.json
n_structure_incomplete_UTF8_BOM.json
i_string_overlong_sequence_6_bytes.json
n_string_invalid-utf-8-in-escape.json
i_string_UTF8_surrogate_U+D800.json
```
Those are all encoding errors and it should not be possible to store those in a Rust String, therefore those are out of scope.

timsueberkrueb committed:
Add comment to transform.py

timsueberkrueb committed:
Update logos to 0.9

DanielBicek committed:
remove unnessessary copying

timsueberkrueb commented:
We're getting there. The `0x00` issue seems to be fixed already in the `v0.10.0` branch for Logos :tada:

timsueberkrueb committed:
Update JSON example for YALR changes

timsueberkrueb committed:
Update logos to 0.10.0-rc1

timsueberkrueb commented:
```
---- n_multidigit_number_then_00 stdout ----
thread 'n_multidigit_number_then_00' panicked at 'assertion failed: Parser::parse_str(r#"123"#).is_err()', examples/json/tests/parsing.rs:781:5
note: Run with `RUST_BACKTRACE=1` for a backtrace.

---- n_string_unescaped_crtl_char stdout ----
thread 'n_string_unescaped_crtl_char' panicked at 'assertion failed: Parser::parse_str(r#"["aa"]"#).is_err()', examples/json/tests/parsing.rs:832:5

---- n_structure_null_dash_byte_dash_outside_dash_string stdout ----
thread 'n_structure_null_dash_byte_dash_outside_dash_string' panicked at 'assertion failed: Parser::parse_str(r#"[]"#).is_err()', examples/json/tests/parsing.rs:661:5

---- n_structure_whitespace_formfeed stdout ----
thread 'n_structure_whitespace_formfeed' panicked at 'assertion failed: Parser::parse_str(r#"[
              ]"#).is_err()', examples/json/tests/parsing.rs:751:5
```

Still failing:
* [x] `n_multidigit_number_then_00` (fixed by specifying trivia)
* [x] `n_structure_whitespace_formfeed` (fixed by specifying trivia)

New failures:
* [x] `n_string_unescaped_crtl_char:
  * parse result: `Ok(Array([String("a\u{0}a")]))` (expected err)
* [x] `n_structure_null_dash_byte_dash_outside_dash_string` (fixed by specifying trivia)

timsueberkrueb commented:
Hey @maciejhirsz, first of all, thanks so much for all your work on `v0.10.0`. 
I've got one question: Is the "old" `0x0`-Byte behavior active when using `WithSource` instead of `BinarySource` in Logos `0.10.0-rc1`? This would explain the `n_multidigit_number_then_00`  issue above. 
EDIT: Okay, 3/4 have been fixed by specifying default trivia. Still, if one would not specify default trivia, `0x0`-bytes might lead to unexpected behavior (that is parsing stopping and potentially succeeding even if it should fail), right? Or am I missing something?

timsueberkrueb committed:
Specify default trivia in JSON example

timsueberkrueb committed:
Forbid control characters in JSON strings
As per RFC 7159, the control characters (U+0000 through U+001F) must be
escaped in strings.

timsueberkrueb commented:
I'm pretty sure that I was wrong with my assumption from yesterday. Looks like the issue was that we were falsely allowing control characters including `\x00` in our JSON strings.
The test is passing now. Sorry for the noise and thanks so much for Logos (and `v0.10.0` in particular) :)

timsueberkrueb committed:
Satisfy clippy and rustfmt

timsueberkrueb reviewed (approved):
Thank you @DanielBicek!
